File: 933. Number of Recent Calls.js
var RecentCounter = function() {
    this.calls = [];
};

/** 
 * @param {number} t
 * @return {number}
 */
RecentCounter.prototype.ping = function(t) {
    
    let diff = (t-3000 >= 0) ? t-3000 : 0;
    let length = this.calls.length;
    
    let start = 0;
    let end = length-1;
    while(start <= end){
        let middle = Math.floor((start+end) /2);
    
        if(diff > this.calls[middle]){
             start = middle+1;
             continue;
        }
        end = middle-1;
      
    }
    this.calls.push(t);
    this.calls.splice(0, start);
    return this.calls.length;
};

File: java/933. Number of Recent Calls.java
class RecentCounter {
    List<Integer> calls;
    public RecentCounter() {
        calls = new ArrayList();
    }
    
    public int ping(int t) {
        int count = 0;
        int length = calls.size();
        int end = length-1;
        int start = 0;
        int diff = (t-3000 >= 0) ? t-3000 : 0;
    
        while(start <= end){
            int middle = (start+end)/2;
            if(diff > calls.get(middle)){
                start = middle+1;
                continue;
            }
                end = middle-1;
        }
        calls.add(t);
        return length +1 - start;
    }
}
